Add more contexts to painless execute api

martijnvg opened the pull request:
This change adds two contexts the execute scripts against:

* SCORE_SCRIPT: Allows to run scripts in a score script context.
This context is used in `function_score` query's script function,

* FILTER_SCRIPT: Allows to run scripts in a filter script context.
This context is used in the `script` query.

In both contexts a index name needs to be specified and a sample document.
The document is needed to create an in-memory index that the script can
access via the `doc[...]` and other notations. The index name is needed
because a mapping is needed to index the document.

Examples:

```
POST /_scripts/painless/_execute
{
  "script": {
    "source": "doc['field'].value.length()"
  },
  "execute_context" : {
    "score": {
      "document": {
        "field": "four"
      },
      "index": "my-index"
    }
  }
}
```

Returns:

```
{
  "result": 4
}
```

```
POST /_scripts/painless/_execute
{
  "script": {
    "source": "doc['field'].value.length() <= params.max_length",
    "params": {
      "max_length": 4
    }
  },
  "execute_context" : {
    "filter": {
      "document": {
        "field": "four"
      },
      "index": "my-index"
    }
  }
}
```

Returns:

```
{
  "result": true
}
```

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented:
> SEARCH_SCRIPT: Allows to run scripts in a search script context.
This context is used in function_score query's script function,
script fields, script sorting and terms_set query.

I think we should have separate context names for script fields and script sorting before this PR. This should be pretty quick to do (see SearchScript.AGGS_CONTEXT as an example) and I'm happy to do the work myself if you don't feel comfortable. All of these cases will eventually change to have their own Script classes, and I want there to be less pain in transitioning to those new contexts in the code.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -86,31 +111,76 @@ public Response newResponse() {
+                    token = p.nextToken();
+                    assert token == XContentParser.Token.END_OBJECT;
+                } else if (supportedContext == SupportedContext.FILTER_SCRIPT || supportedContext == SupportedContext.SEARCH_SCRIPT) {
+                    for (token = p.nextToken(); token != XContentParser.Token.END_OBJECT; token = p.nextToken()) {
```
Can you add an example, preferably in the docs for this, of what the input will look like here? It is difficult to figure out from the code. I also wonder if it would be easier to use object parser internally here?

martijnvg replied:
I still need to add docs, but it going to look like the examples I provided in the PR description:

```
"search_script": {
      "document": {
        "field": "four"
      },
      "index": "my-index"
    }
```

In this context case, a document and an index name needs to be provided.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -86,31 +111,76 @@ public Response newResponse() {
-            return PARSER.parse(parser, null);
+            ParseContext parseContext = new ParseContext();
+            Request request = PARSER.parse(parser, parseContext);
+            request.setIndex(parseContext.index);
```
Isn't the idea of using ObjectParser so the object returned from the parser is completely constructed?

martijnvg replied:
That is true, but that couldn't be done in this case. In the json the document / index name are provided under the context field, which maps the `context` field in the request class, but that is an enum. So via this way we attach the index name and document to the parse context, so that we later apply them to the `document` and `index` fields in the request class.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -193,6 +318,10 @@ public static SupportedContext fromId(byte id) {
                 switch (id) {
                     case 0:
                         return PAINLESS_TEST;
```
We already have a unique identifier for each context (the name). Could we use that instead of creating an enumeration here that will need to be updated for every new context?

martijnvg replied:
Good point, let me try that.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -283,30 +412,89 @@ public PainlessTestScript(Map<String, Object> params) {
+                            SearchScript.LeafFactory leafFactory =
+                                factory.newFactory(request.getScript().getParams(), context.lookup());
+                            SearchScript searchScript = leafFactory.newInstance(leafReaderContext);
+                            Object result = searchScript.run();
```
This is another reason I would like script fields and sort scripts separated out. Actual scoring scripts should call the method expecting a double here.

martijnvg commented:
Thanks for taking a look @rjernst.

> I think we should have separate context names for script fields and script sorting before this PR. This should be pretty quick to do (see SearchScript.AGGS_CONTEXT as an example) and I'm happy to do the work myself if you don't feel comfortable. All of these cases will eventually change to have their own Script classes, and I want there to be less pain in transitioning to those new contexts in the code.

I see, makes sense. I'm happy to try it out myself in separate PRs.

martijnvg commented:
@rjernst I've updated the PR.

Besides addressing your comment, I also made the following changes:
* After #30721 got merged, I exposed SCORE_SCRIPT context instead of SEARCH_SCRIPT context.
* Execute on a management thread instead of a network thread. Not sure if management is the right threadpool, but executing on a network thread is not good anymore, because now also documents get indexed into an in-memory index and I don't consider that a light weight operation and therefor imo can't execute on a network thread.
* Changed the painless execute api's transport action to extend from `TransportSingleShardAction` instead of `HandledTransportAction`. In the case that script or filter script contexts are used then an `IndexService` instance is needed for indexing. Not all nodes may have an instance for the index referenced in request (only nodes with a shard copy for that index). By extending from `TransportSingleShardAction` requests will be redirected to a node that have an `IndexService` instance.

rjernst commented on docs/painless/painless-execute-script.asciidoc:
```diff
@@ -52,4 +52,92 @@ Response:
+    "params": {
+      "max_rank": 4
+    }
+    "context": {
```
I think it is quite confusing using "context" here, since for scripts, context is something different. It would be really good if we had a different term for this environment the script runs in...

martijnvg replied:
Agreed, I will rename it to `execute_script_context`. (after the ExecuteScriptContext class)

martijnvg replied:
I renamed it to `execute_context` (the script part sounded redundant to me): 539abde7b702ddc800c93fed3be2ba47080aa6d2

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -135,33 +348,45 @@ public ActionRequestValidationException validate() {
             super.readFrom(in);
             script = new Script(in);
-            context = SupportedContext.fromId(in.readByte());
+            if (in.getVersion().onOrBefore(Version.V_6_4_0)) {
```
Isn't the execute action only in 6.x right now? Why would wire compat be necessary when no release includes it yet?

martijnvg replied:
It is also in the 6.3 branch. Unless we want to get this change into 6.3 branch too (while 6.3.0 has not been released) then I think we should have this wire backwards compatibility logic.

martijnvg replied:
This API is marked as experimental, which allows us to make changes to request / response format. I don't know what we exactly guarantee for experimental APIs in a cluster with mixed nodes, but if we don't have any guarantees for that then the wire backwards compatibility logic can be removed.

rjernst commented on modules/lang-painless/src/test/resources/rest-api-spec/test/painless/70_execute_painless_scripts.yml:
```diff
@@ -23,3 +38,53 @@
   - match: { result: "-90" }
+
+---
+"Execute with filter context":
```
Is there any way these could be pure unit tests? At least single node tests?

martijnvg replied:
I will look into this.

martijnvg commented:
@rjernst I've updated the PR.

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -82,40 +119,187 @@ public Response newResponse() {
 
         private static final ParseField SCRIPT_FIELD = new ParseField("script");
-        private static final ParseField CONTEXT_FIELD = new ParseField("context");
+        private static final ParseField EXECUTE_CONTEXT_FIELD = new ParseField("execute_context");
```
I think think this name doesn't convey the meaning. Execute context could easily be a normal script context. But this is really using a script context, but providing some arguments to it. I think perhaps just context is better, but maybe the context name could be separated from a separate argument like `context_setup` that is parsed based on which context is chosen?

rjernst commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessExecuteAction.java:
```diff
@@ -171,31 +396,29 @@ public boolean equals(Object o) {
-                        throw new IllegalArgumentException("unknown context [" + id + "]");
-                }
+        static boolean needDocumentAndIndex(ScriptContext<?> scriptContext) {
+            if (scriptContext == FilterScript.CONTEXT) {
```
This can just be:

```
return scriptContext == FilterScript.CONTEXT || scriptContext == ScoreScript.CONTEXT;
```

martijnvg commented:
Thanks @rjernst! I've updated the PR. I like your idea of introducing a `context_setup` parameter, so I've done that and made the `context` parameter just a string parameter.
